### PR TITLE
Label smoothing in vocab parallel cross entropy

### PR DESCRIPTION
**Pull request by MaximumEntropy**
Signed-off-by: MaximumEntropy <sandeep.subramanian.1@umontreal.ca>

### CODE REVIEW

File: apex/transformer/tensor_parallel/cross_entropy.py [hunk 1]
```diff
@@ -72,15 +72,40 @@ def forward(ctx, vocab_parallel_logits, target):
...
+        )
 
         return loss
 
     @staticmethod
     def backward(ctx, grad_output):
 
         # Retreive tensors from the forward path.
-        softmax, target_mask, masked_target_1d = ctx.saved_tensors
+        softmax, target_mask, masked_target_1d, label_smoothing, vocab_size = ctx.saved_tensors
```
Comment (crcrpar) [inline apex/transformer/tensor_parallel/cross_entropy.py:105]:
```suggestion
        softmax, target_mask, masked_target_1d = ctx.saved_tensors
```

File: apex/transformer/tensor_parallel/cross_entropy.py [hunk 2]
```diff
@@ -72,15 +72,40 @@ def forward(ctx, vocab_parallel_logits, target):
...
+
+        label_smoothing = label_smoothing.item()
+        vocab_size = vocab_size.item()
```
Comment (crcrpar) [inline apex/transformer/tensor_parallel/cross_entropy.py:108]:
```suggestion
        label_smoothing = ctx.label_smoothing
        vocab_size = ctx.vocab_size
```

**Review by crcrpar — COMMENTED** [review]
I wrote some trivial comments but sounds good to me in general

File: apex/transformer/tensor_parallel/cross_entropy.py [hunk 3]
```diff
@@ -81,6 +102,10 @@ def backward(ctx, grad_output):
 
         # Retreive tensors from the forward path.
         softmax, target_mask, masked_target_1d = ctx.saved_tensors
+        label_smoothing, vocab_size = ctx.label_smoothing, ctx.vocab_size
+
+        label_smoothing = label_smoothing.item()
+        vocab_size = vocab_size.item()
```
Comment (crcrpar) [inline apex/transformer/tensor_parallel/cross_entropy.py:108]: These two are Python built-in scalar so `.item()` won't work.

**Review by crcrpar — COMMENTED** [review]
Could you add a test case of label_smoothing enabled to https://github.com/NVIDIA/apex/blob/master/tests/L0/run_transformer/test_cross_entropy.py?

File: apex/transformer/tensor_parallel/cross_entropy.py (hunk 3 above)
Comment (MaximumEntropy) [reply apex/transformer/tensor_parallel/cross_entropy.py:108]: Oh sorry I should have removed those.

**Comment by MaximumEntropy** [comment]
@crcrpar I just added label smoothing args to the cross entropy tests.

**Review by crcrpar — APPROVED** [review]
thank you
